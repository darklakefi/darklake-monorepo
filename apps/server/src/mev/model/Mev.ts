import { IsEnum, IsIn, IsInt, IsOptional, Max, Min, Validate } from "class-validator";
import { IsValidSolanaAddress } from "../../validator/IsValidSolanaAddress";
import { Type } from "class-transformer";
import { SandwichEvent } from "@prisma/client";
import { formatSolAmount } from "src/utils/blockchain";

export class GetMevTotalExtractedQuery {
  @Validate(IsValidSolanaAddress)
  address: string;
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}

export enum MevAttacksOrderBy {
  "DATE" = "date",
  "AMOUNT_DRAINED" = "amountDrained",
}

export class GetMevAttacksQuery {
  @Validate(IsValidSolanaAddress)
  address: string;

  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(100)
  @Type(() => Number)
  limit?: number;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  offset?: number;

  @IsOptional()
  @IsEnum(MevAttacksOrderBy)
  orderBy?: MevAttacksOrderBy;

  @IsOptional()
  @IsIn(["asc", "desc"])
  direction?: "asc" | "desc";
}

export interface Mev {
  tokenName: string;
  timestamp: number;
  solAmountSent: number;
  solAmountDrained: number;
  transactionHash: string;
}

export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export class MevAttackResponse {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };

  constructor(props: SandwichEvent) {
    this.solAmount = {
      sent: formatSolAmount(props.sol_amount_swap),
      lost: formatSolAmount(props.sol_amount_drained),
    };

    this.tokenName = "UNKNOWN"; // TODO: add token metadata service & get token name
    this.timestamp = +props.occurred_at;
    this.transactions = {
      frontRun: {
        address: "0x0", // TODO: add address
        signature: "0x0",
      },
      victim: {
        address: "0x0",
        signature: props.tx_hash_victim_swap,
      },
      backRun: {
        address: "0x0",
        signature: "0x0",
      },
    };
  }
}
