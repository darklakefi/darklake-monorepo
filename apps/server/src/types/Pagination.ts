export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum PaginatedResponseDataLimit {
  MEV_EVENTS_LIST = 100,
}
