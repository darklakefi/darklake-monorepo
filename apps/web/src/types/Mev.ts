export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export interface MevAttack {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}
