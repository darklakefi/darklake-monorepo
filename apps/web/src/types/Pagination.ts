export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum SortDirection {
  ASC = "asc",
  DESC = "desc",
}
