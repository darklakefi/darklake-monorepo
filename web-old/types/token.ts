export interface Token {
  symbol: string;
  name: string;
  image: string;
  address: string;
  tokenProgram: string;
  decimals: number;
}
