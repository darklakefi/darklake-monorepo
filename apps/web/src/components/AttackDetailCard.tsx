import ProgressBar from "@/components/ProgressBar";
import { MevAttack } from "@/types/Mev";
import { format } from "date-fns";
import { ReactNode } from "react";

export interface AttackDetailCardProps {
  mevAttack: MevAttack;
  cardTitle?: string | ReactNode;
  onOpenModal?: () => void;
}

const AttackDetailCard = ({ mevAttack, cardTitle, onOpenModal }: AttackDetailCardProps) => {
  const handleOpenBreakdownModal = () => {
    if (onOpenModal) {
      onOpenModal();
    }
  };

  const extractedPercentageRounded = Number(((mevAttack.solAmount.lost / mevAttack.solAmount.sent) * 100).toFixed(2));
  return (
    <div className="flex flex-col bg-brand-60 gap-4 p-4 h-full">
      <div className="flex flex-col">
        {cardTitle && <div className="text-lg leading-6 tracking-normal text-brand-20">{cardTitle}</div>}
        <div className="text-lg leading-6 tracking-normal text-brand-30 uppercase">TOKEN: {mevAttack.tokenName}</div>
        <div className="text-lg leading-6 tracking-normal text-brand-30">
          {format(mevAttack.timestamp || new Date(), "yyyy-MM-dd HH:mm 'UTC'")}
        </div>
      </div>
      <div className="flex flex-col gap-3 py-4 border-t border-b border-brand-40">
        <div className="text-3xl leading-8 tracking-wide">{mevAttack.solAmount.lost} SOL LOST</div>
        <div className="text-lg leading-6 tracking-normal">
          <ProgressBar progress={extractedPercentageRounded} />
        </div>
        <div className="flex flex-col">
          <div className="text-lg leading-6 tracking-normal text-brand-20 uppercase">
            {`${extractedPercentageRounded}% EXTRACTED`}
          </div>
          <div className="text-lg leading-6 tracking-normal text-brand-30">
            FROM A {mevAttack.solAmount.sent} SOL TRANSACTION
          </div>
        </div>
      </div>
      <button
        className="uppercase underline text-lg leading-6 tracking-normal text-brand-30 w-fit"
        onClick={handleOpenBreakdownModal}
      >
        View attack breakdown
      </button>
    </div>
  );
};

const BlurredMode = ({ index }: { index: number }) => {
  return (
    <div className="flex flex-col bg-brand-60 gap-4 p-4 h-full">
      <div className="flex flex-col">
        <div className="text-lg leading-6 tracking-normal text-brand-20">#{index} LARGEST EXTRACTION EVENT</div>
        <div className="text-lg leading-6 tracking-normal text-brand-30">TOKEN: █████</div>
        <div className="text-lg leading-6 tracking-normal text-brand-30">████-██-██ ██:██ UTC</div>
      </div>
      <div className="flex flex-col gap-3 py-4 border-t border-b border-brand-40">
        <div className="text-3xl leading-8 tracking-wide">██.██ SOL LOST</div>
        <div className="text-lg leading-6 tracking-normal">░░░░░░░░░░░░░░░░░░░░░░░░░░</div>
        <div className="flex flex-col">
          <div className="text-lg leading-6 tracking-normal text-brand-20">██% EXTRACTED</div>
          <div className="text-lg leading-6 tracking-normal text-brand-30">FROM A ██.██ SOL TRANSACTION</div>
        </div>
      </div>
      <button className="uppercase underline text-lg leading-6 tracking-normal text-brand-30 w-fit">
        View attack breakdown
      </button>
    </div>
  );
};

AttackDetailCard.Blurred = BlurredMode;

export default AttackDetailCard;
