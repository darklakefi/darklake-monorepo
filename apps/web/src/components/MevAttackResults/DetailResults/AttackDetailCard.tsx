import ProgressBar from "@/components/ProgressBar";
import { MevAttack } from "@/types/Mev";
import { format } from "date-fns";

type AttackDetailCardProps = { mevAttack: MevAttack; index?: number; onOpenModal?: () => void };

const AttackDetailCard = ({ mevAttack, index, onOpenModal }: AttackDetailCardProps) => {
  const handleOpenBreakdownModal = () => {
    if (onOpenModal) {
      onOpenModal();
    }
  };

  return (
    <div className="flex flex-col bg-brand-60 gap-[16px] p-[16px] h-full">
      <div className="flex flex-col">
        {index && <div className="text-body-2 text-brand-20">#{index} LARGEST EXTRACTION EVENT</div>}
        <div className="text-body-2 text-brand-30">TOKEN: {mevAttack.tokenName}</div>
        <div className="text-body-2 text-brand-30">
          {format(mevAttack.timestamp || new Date(), "yyyy-MM-dd HH:mm 'UTC'")}
        </div>
      </div>
      <div className="flex flex-col gap-[12px] py-[16px] border-t border-b border-brand-40">
        <div className="text-body">{mevAttack.solAmount.lost} SOL LOST</div>
        <div className="text-body-2">
          <ProgressBar progress={(mevAttack.solAmount.lost / mevAttack.solAmount.sent) * 100} />
        </div>
        <div className="flex flex-col">
          <div className="text-body-2 text-brand-20">64% EXTRACTED</div>
          <div className="text-body-2 text-brand-30">FROM A {mevAttack.solAmount.sent} SOL TRANSACTION</div>
        </div>
      </div>
      <button className="text-link text-brand-30 w-fit" onClick={handleOpenBreakdownModal}>
        View attack breakdown
      </button>
    </div>
  );
};

const BlurredMode = ({ index }: { index: number }) => {
  return (
    <div className="flex flex-col bg-brand-60 gap-[16px] p-[16px] h-full">
      <div className="flex flex-col">
        <div className="text-body-2 text-brand-20">#{index} LARGEST EXTRACTION EVENT</div>
        <div className="text-body-2 text-brand-30">TOKEN: █████</div>
        <div className="text-body-2 text-brand-30">████-██-██ ██:██ UTC</div>
      </div>
      <div className="flex flex-col gap-[12px] py-[16px] border-t border-b border-brand-40">
        <div className="text-body">██.██ SOL LOST</div>
        <div className="text-body-2">░░░░░░░░░░░░░░░░░░░░░░░░░░</div>
        <div className="flex flex-col">
          <div className="text-body-2 text-brand-20">██% EXTRACTED</div>
          <div className="text-body-2 text-brand-30">FROM A ██.██ SOL TRANSACTION</div>
        </div>
      </div>
      <button className="text-link text-brand-30 w-fit" onClick={() => {}}>
        View attack breakdown
      </button>
    </div>
  );
};

AttackDetailCard.Blurred = BlurredMode;

export default AttackDetailCard;
