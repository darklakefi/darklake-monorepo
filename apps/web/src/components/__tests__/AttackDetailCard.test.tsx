import { act, render, RenderResult } from "@testing-library/react";
import { MevAttack, MevAttackSwapType, MevTransaction } from "@/types/Mev";
import AttackDetailCard from "../AttackDetailCard";

const mockMevAttackTransaction: MevTransaction = {
  address: "someAddress",
  signature: "someSignature",
};

const mockMevAttack: MevAttack = {
  swapType: MevAttackSwapType.BUY,
  timestamp: +new Date("2024-03-26"),
  tokenName: "TRUMP",
  solAmount: {
    sent: 100,
    lost: 30,
  },
  transactions: {
    frontRun: mockMevAttackTransaction,
    victim: mockMevAttackTransaction,
    backRun: mockMevAttackTransaction,
  },
};

describe("<AttackDetailCard />", () => {
  it("displays attack detail", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AttackDetailCard mevAttack={mockMevAttack} />));
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("display attack detail with top index", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AttackDetailCard mevAttack={mockMevAttack} />));
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("displays blurred attack detail", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AttackDetailCard.Blurred index={1} />));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
