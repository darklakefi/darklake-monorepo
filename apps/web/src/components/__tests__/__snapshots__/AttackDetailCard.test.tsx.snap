// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttackDetailCard /> display attack detail with top index 1`] = `
<body>
  <div>
    <div
      class="flex flex-col bg-brand-60 gap-4 p-4 h-full"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="text-lg leading-6 text-brand-30"
        >
          TOKEN:
          TRUMP
        </div>
        <div
          class="text-lg leading-6 text-brand-30"
        >
          2024-03-26 00:00 UTC
        </div>
      </div>
      <div
        class="flex flex-col gap-3 py-4 border-t border-b border-brand-40"
      >
        <div
          class="text-3xl leading-8"
        >
          30
           SOL LOST
        </div>
        <div
          class="text-lg leading-6"
        >
          <div
            class="text-lg flex flex-row overflow-hidden select-none"
          >
            <div
              class="overflow-hidden tracking-[1px]"
              style="width: 30%;"
            >
              ██████████████████████████████
            </div>
            <div
              class="overflow-hidden"
              style="width: 70%;"
            >
              ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
            </div>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="text-lg leading-6 text-brand-20"
          >
            64% EXTRACTED
          </div>
          <div
            class="text-lg leading-6 text-brand-30"
          >
            FROM A
            100
             SOL TRANSACTION
          </div>
        </div>
      </div>
      <button
        class="uppercase underline text-lg leading-6 text-brand-30 w-fit"
      >
        View attack breakdown
      </button>
    </div>
  </div>
</body>
`;

exports[`<AttackDetailCard /> displays attack detail 1`] = `
<body>
  <div>
    <div
      class="flex flex-col bg-brand-60 gap-4 p-4 h-full"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="text-lg leading-6 text-brand-30"
        >
          TOKEN:
          TRUMP
        </div>
        <div
          class="text-lg leading-6 text-brand-30"
        >
          2024-03-26 00:00 UTC
        </div>
      </div>
      <div
        class="flex flex-col gap-3 py-4 border-t border-b border-brand-40"
      >
        <div
          class="text-3xl leading-8"
        >
          30
           SOL LOST
        </div>
        <div
          class="text-lg leading-6"
        >
          <div
            class="text-lg flex flex-row overflow-hidden select-none"
          >
            <div
              class="overflow-hidden tracking-[1px]"
              style="width: 30%;"
            >
              ██████████████████████████████
            </div>
            <div
              class="overflow-hidden"
              style="width: 70%;"
            >
              ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
            </div>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="text-lg leading-6 text-brand-20"
          >
            64% EXTRACTED
          </div>
          <div
            class="text-lg leading-6 text-brand-30"
          >
            FROM A
            100
             SOL TRANSACTION
          </div>
        </div>
      </div>
      <button
        class="uppercase underline text-lg leading-6 text-brand-30 w-fit"
      >
        View attack breakdown
      </button>
    </div>
  </div>
</body>
`;

exports[`<AttackDetailCard /> displays blurred attack detail 1`] = `
<body>
  <div>
    <div
      class="flex flex-col bg-brand-60 gap-4 p-4 h-full"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="text-lg leading-6 text-brand-20"
        >
          #
          1
           LARGEST EXTRACTION EVENT
        </div>
        <div
          class="text-lg leading-6 text-brand-30"
        >
          TOKEN: █████
        </div>
        <div
          class="text-lg leading-6 text-brand-30"
        >
          ████-██-██ ██:██ UTC
        </div>
      </div>
      <div
        class="flex flex-col gap-3 py-4 border-t border-b border-brand-40"
      >
        <div
          class="text-3xl leading-8"
        >
          ██.██ SOL LOST
        </div>
        <div
          class="text-lg leading-6"
        >
          ░░░░░░░░░░░░░░░░░░░░░░░░░░
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="text-lg leading-6 text-brand-20"
          >
            ██% EXTRACTED
          </div>
          <div
            class="text-lg leading-6 text-brand-30"
          >
            FROM A ██.██ SOL TRANSACTION
          </div>
        </div>
      </div>
      <button
        class="uppercase underline text-lg leading-6 text-brand-30 w-fit"
      >
        View attack breakdown
      </button>
    </div>
  </div>
</body>
`;
